File: cpp/60_把二叉树打印成多行.cpp
#include <vector>
#include <queue>
#include <stdio.h>
#include <sys/time.h>

#include "TreeNode.h"
#include "my_vector.h"

/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/

class Solution {
public:
    // 1. 两个向量交替存储
    std::vector<std::vector<int> > PrintLevelOrder1(TreeNode* root) {
        if (root == nullptr) {
            return std::vector<std::vector<int>>{};
        }

        std::vector<std::vector<int>> ret;
        std::vector<TreeNode*> curr_list = {root};
        
        while (!curr_list.empty()) {
            std::vector<int> curr_vec;
            std::vector<TreeNode*> next_list;
            for (auto & i : curr_list) {
                curr_vec.push_back(i->val);
                if (i->left != nullptr) {
                    next_list.push_back(i->left);
                }
                if (i->right != nullptr) {
                    next_list.push_back(i->right);
                }
            }
            ret.push_back(curr_vec);
            curr_list = next_list;
        }
        
        return ret;
    }

    // 2. 队列法，bfs
    std::vector<std::vector<int>> PrintLevelOrder2(TreeNode* root) {
        if (root == nullptr) {
            return std::vector<std::vector<int>>{};
        }

        std::vector<std::vector<int>> ret;
        std::queue<TreeNode*> queue;
        queue.push(root);

        while (!queue.empty()) {
            std::vector<int> curr_vec;
            int last_num = queue.size();
            for (int i = 0; i < last_num; i++) { // 遍历提取上一层
                TreeNode* node = queue.front();
                queue.pop();
                curr_vec.push_back(node->val);
                if (node->left) {
                    queue.push(node->left);
                }
                if (node->right) {
                    queue.push(node->right);
                }
            }
            ret.push_back(curr_vec);
        }

        return ret;
    }

    // 3. 递归法，dfs
    std::vector<std::vector<int>> PrintLevelOrder3(TreeNode* root) {
        std::vector<std::vector<int>> ret;
        dfs(root, ret, 0);
        return ret;
    }

    void dfs(TreeNode* root, std::vector<std::vector<int>>& ret, int level) {
        if (root == nullptr) {
            return;
        }
        if (level >= ret.size()) {
            ret.push_back(std::vector<int>{});
        }
        ret[level].push_back(root->val);
        dfs(root->left, ret, level + 1);
        dfs(root->right, ret, level + 1);
        return;
    }

    // 测试函数
    void test(std::vector<int> vec) {
        std::vector<std::vector<int>> result;
        struct timeval start, end;
        TreeNode* head = TreeNode_creatFromFrontOrder(vec);
        printf("=====\n");
        for (auto & func : this->func_vec_) {
            gettimeofday(&start, nullptr);
            result = (this->*func)(head);
            gettimeofday(&end, nullptr);
            printf("times(us): %ld, result: \n", end.tv_usec - start.tv_usec);
            printf_2d_vec(result);
        }
    }

private:
    typedef std::vector<std::vector<int>> (Solution::*func_ptr)(TreeNode*);
    std::vector<func_ptr> func_vec_ = {&Solution::PrintLevelOrder1,
                                       &Solution::PrintLevelOrder2,
                                       &Solution::PrintLevelOrder3};
};

int main(int argc, char* argv[])
{
    std::vector<int> vec = {1, 2, 4, -1, -1, 5, -1, -1, 3, 6, -1, -1, 7, -1, -1};
    std::vector<int> vec2 = {3, 9, -1, -1, 20, 15, -1, -1, 7, -1, -1};

    Solution s;
    s.test(vec);
    s.test(vec2);

    return 0;
}

File: sword_offer/python/60_把二叉树打印成多行.py
#!/bin/bash/ python3
#-*- coding:utf-8 -*-

import time
import TreeNode

"""
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
"""

class Solution():
    def PrintLevelOrder1(self, root):
        """
        1. 两个向量交替存储
        """
        if root == None:
            return []
        
        ret, curr_list = [], [root]

        while curr_list:
            curr_vec = []
            next_list = []
            for i in curr_list:
                curr_vec.append(i.val)
                if i.left:
                    next_list.append(i.left)
                if i.right:
                    next_list.append(i.right)
            ret.append(curr_vec)
            curr_list = next_list
            
        return ret
    
    def PrintLevelOrder2(self, root):
        """
        2. 队列法，bfs
        """
        if root == None:
            return
        
        ret, queue = [], [root]

        while queue:
            curr_vec = []
            for _ in range(len(queue)):
                node = queue.pop(0)
                curr_vec.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            ret.append(curr_vec)
        
        return ret
    
    def PrintLevelOrder3(self, root):
        """
        3. 递归法，dfs
        """
        def dfs(root, ret, level):
            if root == None:
                return
            if level >= len(ret):
                ret.append([])
            ret[level].append(root.val)
            dfs(root.left, ret, level + 1)
            dfs(root.right, ret, level + 1)
            return
        
        ret = []
        dfs(root, ret, 0)
        return ret
    
    def test(self, vec):
        """
        测试函数
        """
        func_vec = [self.PrintLevelOrder1,
                    self.PrintLevelOrder2,
                    self.PrintLevelOrder3]
        root = TreeNode.TreeNode().creatFromFrontOrder(vec)
        print("=====")
        for func in func_vec:
            start = time.time()
            result = func(root)
            end = time.time()
            print("times(us): {:>5.2f}, result: {}".format((end - start)*10**6, result))
        

def main():
    vec = [1, 2, 4, -1, -1, 5, -1, -1, 3, 6, -1, -1, 7, -1, -1]
    vec2 = [3, 9, -1, -1, 20, 15, -1, -1, 7, -1, -1]

    s = Solution()
    s.test(vec)
    s.test(vec2)


if __name__ == "__main__":
    main()